### PR TITLE
Fuse fixes

### PR DESCRIPTION
**Pull request by jdoliner**

### CODE REVIEW

**Comment by jdoliner** [comment]
The high level summary of these changes is that they make the fuse layer much faster. They do this by sending a single request for the object, writing it out to a file and then using that file to server later read requests. Previously each read request would be another request to Pachyderm which was very slow because the OS issued a request for each 128K chunk of the file, so even a decently size file would result in many many requests.

File: src/server/pfs/fuse/file.go [hunk 1]
```diff
@@ -38,18 +83,19 @@ func (f *file) InnerFile() nodefs.File {
...
-	switch {
-	case repo != nil:
-		return nil, fuse.Status(syscall.EISDIR)
-	case file != nil:
-		return newReadResult(f.fs.c, file, offset, len(dest)), fuse.OK
-	default:
-		return nil, fuse.Status(syscall.EISDIR)
+	f.counter.wait(waitn)
+	// check if there was an error reading the file
+	if f.err != nil {
```
Comment (sjezewski) [inline src/server/pfs/fuse/file.go]: So ... if the file was created w a call to `newFile()` but FUSE never receives any reads (presumably calls to this `Read` function), then we never report the error? Does it make sense to have a logger emit this error right when it happens (after the `GetFile`?)

File: src/server/pfs/fuse/filesystem.go [hunk 2]
```diff
@@ -204,6 +204,7 @@ func (fs *filesystem) fileAttr(f *pfs.File) (*fuse.Attr, fuse.Status) {
 	if err != nil {
 		return nil, toStatus(err)
 	}
+	fmt.Printf("Size: %d\n", fi.SizeBytes)
```
Comment (sjezewski) [inline src/server/pfs/fuse/filesystem.go]: Is this a debug line?

File: src/server/pfs/fuse/fuse_test.go [hunk 3]
```diff
@@ -40,18 +44,48 @@ func TestBasic(t *testing.T) {
...
+		require.NoError(t, err)
+		require.Equal(t, int(pfs.ChunkSize), len(data))
+	})
+}
+
 func TestLargeFile(t *testing.T) {
 	c := server.GetPachClient(t)
 	require.NoError(t, c.CreateRepo("repo"))
-	_, err := c.PutFile("repo", "master", "file", strings.NewReader(strings.Repeat("p", GB)))
+	src := workload.RandString(rand.New(rand.NewSource(123)), GB+17)
```
Comment (sjezewski) [inline src/server/pfs/fuse/fuse_test.go]: Just curious about the GB threshold ... I'm guessing you wanted to test something just past 1GB and the `17` is an arbitrary small number to get past 1GB

Comment (jdoliner) [reply src/server/pfs/fuse/fuse_test.go]: Yup that's the intention here. FUSE sends requests for 128K chunks so I wanted something not divisible by that so we'd get some differently sized requests at the end.

File: src/server/pfs/fuse/filesystem.go (hunk 2 above)
Comment (jdoliner) [reply src/server/pfs/fuse/filesystem.go]: Yup, good catch, will remove.

File: src/server/pfs/fuse/fuse_test.go (hunk 3 above)
Comment (sjezewski) [reply src/server/pfs/fuse/fuse_test.go]: :+1:

File: src/server/pfs/fuse/file.go (hunk 1 above)
Comment (sjezewski) [reply src/server/pfs/fuse/file.go]: Fair enough. I can see how it'd generate noise before a user actually attempts a read.

Comment (jdoliner) [reply src/server/pfs/fuse/file.go]: Yeah, that's my main worry. The FUSE mount is often backgrounded within a shell, so if it gets too chatty it makes the shell unusable.

**Pull request merged**